### ISSUE TITLE
Exported ambient variables are neither emitted nor an error

### ISSUE THREAD

**Issue by Arnavion** [microsoft/TypeScript#8015]
**TypeScript Version:** 

1.9.0-dev.20160411

**Code**

foo.ts:

``` typescript
export type Foo = { bar: string; }
declare var foo: Foo;
export { foo };
```

bar.ts:

``` typescript
import { foo } from "./foo";
console.log(foo.bar);
```

`tsc -t es5 -m commonjs ./foo.ts ./bar.ts`

**Expected behavior:**

Either it should be an error to export an ambient variable, or it should be emitted something like this:

foo.js:

``` javascript
"use strict";
exports.foo = foo;
```

bar.js:

``` javascript
"use strict";
var foo_1 = require("./foo");
console.log(foo_1.foo.bar);
```

**Actual behavior:** 

No error, and foo.js is emitted with no export, causing bar.js to blow up at runtime.

foo.js:

``` javascript
"use strict";
```

**Comment by falsandtru** [comment]
Expected behavior of `foo.js` always throws ReferenceError.

```js
"use strict";
exports.foo = foo; // foo is not defined, also cannot reference global scope.
```

It should be

```js
"use strict";
```

So I think, actual behavior is correct.

**Comment by falsandtru** [comment]
It only works as a type declaration of foo module. So I think, actual behavior is correct. If it will be compile errors, TypeScript disallows exporting of types in the variable namespace. 

```ts
import { foo } from "./foo";
var bar: typeof foo;
```

**Comment by Arnavion** [comment]
@falsandtru I'm not sure what the ES6 spec says, but the current behavior of TS is NOT correct. It accepts the code in the OP without error and creates output that throws error at runtime. Either it needs to consider the code an error, or it needs to create output that works. Hence this bug report.

**Comment by falsandtru** [comment]
@Arnavion I could not get an error in my example code. Can you give me an example code of your actual use case?

**Comment by mhegazy** [comment]
Just to clarify, the ES6 spec specifies that an export has to have a local binding. so 

```ts
export {global};
```

is illegal.

in the OP, the binding is local, which should be fine. it just happened to be ambient. so it should be have like:

```
var local: number;
export {local};
```

**Comment by falsandtru** [comment]
@mhegazy if you fix this issue, **it will be a breaking change** by disallowing variable type declaration. For example, the following code throws no error currently, but it will throw an error after fix this issue.

```ts
// bar.ts
import "./foo";
import { foo } from "./foo";
var bar: typeof foo;
```

```js
// bar.js
require("./foo"); // get an error at `exports.foo = foo;`
var bar;
```

**Comment by Arnavion** [comment]
@mhegazy Yeah, I did think `export` would only allow local bindings, so yes TS should make that an error.

The intent was to use an external global and annotate it with a local type, so the workaround I went with was:

```typescript
export type Foo = { bar: string; }
declare var foo: Foo;
var globalFoo = foo; // local binding to global
export { globalFoo as foo };
```

**Comment by falsandtru** [comment]
@Arnavion your new code does not export `foo`. It code does not need to fix by the compiler. If you want to fix

from

```ts
export type Foo = { bar: string; }
declare var foo: Foo;
var globalFoo = foo; // local binding to global
export { globalFoo as foo };
```

to

```ts
export type Foo = { bar: string; }
declare var foo: Foo; // local binding to global
export { foo };
```

, it is illegal.

**Comment by ddotlic** [comment]
@Arnavion Hey, I just had exact same problem!

> The intent was to use an external global and annotate it with a local type

Exactly what I tried to achieve.

I ended up doing (this satisfies the compiler and emits nothing) the following
`export declare var foo : Foo`

This is all in a `.ts` file and not in a `.d.ts`, just to be clear.

Frankly, I was tired of trying various things and desperately typed this and it worked, ashamed to say I don't know why; TypeScript 1.8.9

**Comment by ddotlic** [comment]
Forgot to say, this is for the browser, so the fact that nothing is emitted suites me fine, the variable `foo` will be provided by "ambient" JavaScript (which will simply hang it off of the global `window`).

If it was node, this indeed would have blown up too.

**Comment by mhegazy** [comment]
Just to clarify:

Exporting a name that was **not** declared in the module is an error as per the ES6 spec. the compiler flags this correctly today. see https://github.com/Microsoft/TypeScript/issues/6323#issuecomment-170170230

Exporting a name that **is** declared in the module is legal, regardless if it is ambient or not, the emitter should emit it correctly, and this issue is tracking fixing this.
so 
```ts
declare var foo: Foo;
export { foo };
```
should emit
```js
exports.foo = foo; 
```

Now, one thing to note, if you are doing this to capture a global variable, which it seems @Arnavion and @ddotlic are doing, then you need to be aware that this is going to be an error if/when your run this code under a native ES6 module implementation. the fact that the compiler does not flag it does not make it ok at runtime. so consider declaring a local name:

```ts
var globalWindow = window;
export {globalWindow as window};
```

**Comment by Arnavion** [comment]
Yes, that is the workaround I posted.

**Comment by RyanCavanaugh** [comment]
Confirmed fixed
